Fix fork io testing

hhefesto opened the pull request:
Fixes `cabal test`

Now telomare runs can a priori be given a list of inputs that'll be fed
when input is required. Also the output of telomare runs can now be
captured which is nice for testing.

hhefesto committed:
Code taken from 776e2d7 by sfultong

hhefesto committed:
Cleanup app/Main.hs unused imports

hhefesto committed:
Removed IO fork to get telomare run output

Now telomare runs can a priori be given a list of inputs that'll be fed
when input is required. Also the output of telomare runs can now be
captured which is nice for testing.

Also some cleaning.

sfultong commented on src/Telomare/Eval.hs:
```diff
@@ -221,42 +242,58 @@ schemeEval iexpr = do
+evalLoopWithInput :: [String] -> IExpr -> IO String
+evalLoopWithInput inputList iexpr = evalLoopCore iexpr printAcc "" inputList
+  where
+    printAcc acc out = pure (acc <> "\n" <> out)
```
ah, looks like this is the newline issue. It's not a big deal, though

hhefesto replied:
I already notified you in a DM, but for completeness I'll repeat it here:

So this change in `printAcc` makes `cabal test` stall:

```haskell
evalLoopWithInput :: [String] -> IExpr -> IO String
evalLoopWithInput inputList iexpr = evalLoopCore iexpr printAcc "" inputList
  where
    -- printAcc acc out = pure (acc <> "\n" <> out)
    printAcc acc out = if null acc && null out
                       then pure "\n"
                       else if null acc
                            then pure out
                            else pure $ acc <> "\n" <> out
```

The stalling occurs with a simpler `printAcc` too:

```haskell
printAcc acc out = if null acc
                   then pure out
                   else pure (acc <> "\n" <> out)

```

my curiosity has sadly made me think of this more than I thought I would. lol..

hhefesto replied:
The stalling was a symptom to a bigger problem: there was a problem in the `shrink` of `Pattern` (a type that's parsed form the case-patterns to be matched.

Still not sure why changing `printAcc` makes that problem evident (or why it did work with the not so smart version `pure $ acc <> "\n" <> out`).

I also noticed that in my testing `@?=` is used and that it isn't a symmetric operator, the left part is the thing to be tested and the right part is the expected value. This next snippet shows why it matters (expected is inverted with result):

```
  Unit tests on case expressions
    test case with int leaves:          FAIL (0.04s)
      test/CaseTests.hs:73:
      expected: "\nTrue\ndone"
       but got: "True\ndone"
```

sfultong reviewed (commented):
remove debugging, clean up newline issue if you want

I've been thinking: do you want me to give you more stylistic feedback? Sometimes I think I could do things in a cleaner way, but it's fine as-is

hhefesto commented:
> I've been thinking: do you want me to give you more stylistic feedback? Sometimes I think I could do things in a cleaner way, but it's fine as-is

any input is more than welcomed : )

hhefesto commented on src/Telomare/Resolver.hs:
```diff
@@ -31,7 +31,7 @@ import Telomare.Parser (AnnotatedUPT, TelomareParser, parseWithPrelude)
 
 debug :: Bool
-debug = True
+debug = False
```
So should I remove this debug thing you made? Or is it enough with `debug = False` ?

sfultong replied:
just change it to False

hhefesto committed:
Fix `Pattern` `shrink` and beggining newline

Also inverted the arguments of `@?=` where necessary (right side is for
expected value) and did some cleanup
